airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AirbyteFileOffsetBackingStore.java

```diff
@@ -0,0 +1,146 @@
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.source.jdbc;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.integrations.source.jdbc.models.CdcState;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.io.FileUtils;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.storage.FileOffsetBackingStore;
+import org.apache.kafka.connect.util.SafeObjectInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class handles reading and writing a debezium offset file. In many cases it is duplicating
+ * logic in debezium because that logic is not exposed in the public API. We mostly treat the
+ * contents of this state file like a black box. We know it is a Map<ByteBuffer, Bytebuffer>. We
+ * deserialize it to a Map<String, String> so that the state file can be human readable. If we ever
+ * discover that any of the contents of these offset files is not string serializable we will likely
+ * have to drop the human readability support and just base64 encode it.
+ */
+public class AirbyteFileOffsetBackingStore {
+
+  public static final Path DEFAULT_OFFSET_STORAGE_PATH = Path.of("/tmp/offset.dat");
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AirbyteFileOffsetBackingStore.class);
+
+  private final Path offsetFilePath;
+
+  public AirbyteFileOffsetBackingStore() {
+    this(DEFAULT_OFFSET_STORAGE_PATH);
+  }
+
+  @VisibleForTesting
+  AirbyteFileOffsetBackingStore(final Path offsetFilePath) {
+    this.offsetFilePath = offsetFilePath;
+  }
+
+  public CdcState read() {
+    final Map<ByteBuffer, ByteBuffer> raw = load();
+
+    final Map<String, String> mappedAsStrings = raw.entrySet().stream().collect(Collectors.toMap(
+        e -> e.getKey() != null ? new String(e.getKey().array(), StandardCharsets.UTF_8) : null,
+        e -> e.getValue() != null ? new String(e.getValue().array(), StandardCharsets.UTF_8) : null));
+    final JsonNode asJson = Jsons.jsonNode(mappedAsStrings);
+
+    LOGGER.info("debezium state. {}", asJson);
+
+    return new CdcState().withState(asJson);
+  }
+
+  @SuppressWarnings("unchecked")
+  public void persist(CdcState cdcState) {
+    final Map<String, String> mapAsString =
+        cdcState != null && cdcState.getState() != null ? Jsons.object(cdcState.getState(), Map.class) : Collections.emptyMap();
+    final Map<ByteBuffer, ByteBuffer> mappedAsStrings = mapAsString.entrySet().stream().collect(Collectors.toMap(
```

Comment:
Should be DRY-ed (fine if it happens after this PR)

Reply:
fixed.